File: Python/Problems/Problems/problem2210.py
from typing import List

from .problem import Problem


class Problem2210(Problem):
    def test(self) -> bool:
        nums = [2, 4, 1, 1, 6, 5]

        expected = 3

        result = self.countHillValley(nums)

        print(result)

        return result == expected

    def countHillValley(self, nums: List[int]) -> int:
        n = len(nums)
        result = 0
        prev = 0
        i = 0
        while i < n:
            while i < n and nums[i] == nums[prev]:
                i += 1
            if i == n:
                break
            next = i
            while next < n and nums[next] == nums[i]:
                next += 1
            if next == n:
                break
            if ((nums[prev] < nums[i] and nums[next] < nums[i]) or
                    (nums[prev] > nums[i] and nums[next] > nums[i])):
                result += 1
            prev = i
            i = next - 1


        return result

File: C++/Problems/Problems/problem_2210.cpp
#include "problem_2210.h"

bool problem_2210::test() {
    std::vector nums { 2, 4, 1, 1, 6, 5 };

    auto expected = 3;

    auto result = countHillValley(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_2210::countHillValley(std::vector<int> &nums) {
    int n = nums.size();
    auto result = 0;
    auto prev = 0;
    for (int i = 1; i < n - 1; ++i)
    {
        while (i < n && nums[i] == nums[prev])
            i += 1;
        if (i == n)
            break;
        auto next = i;
        while (next < n && nums[next] == nums[i])
            next++;
        if (next == n)
            break;
        if ((nums[prev] < nums[i] && nums[next] < nums[i]) ||
            (nums[prev] > nums[i] && nums[next] > nums[i]))
            result += 1;
        prev = i;
        i = next - 1;
    }

    return result;
}
